File: Delivery.java
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Delivery {
    // Map to store driver information: driver_id -> hourly_rate
    private final Map<Integer, Double> drivers;
    // List to store delivery records
    private final List<DeliveryRecord> deliveries;

    public Delivery() {
        this.drivers = new HashMap<>();
        this.deliveries = new ArrayList<>();
    }

    /**
     * Add a new driver with their hourly rate.
     *
     * @param driverId Unique identifier for the driver
     * @param usdHourlyRate Driver's hourly rate in USD
     * @throws IllegalArgumentException if driver already exists or rate is invalid
     */
    public void addDriver(int driverId, double usdHourlyRate) {
        if (drivers.containsKey(driverId)) {
            throw new IllegalArgumentException("Driver with ID " + driverId + " already exists");
        }
        if (usdHourlyRate <= 0) {
            throw new IllegalArgumentException("Hourly rate must be greater than 0");
        }
        
        drivers.put(driverId, usdHourlyRate);
    }

    /**
     * Record a delivery with its start and end times.
     *
     * @param driverId ID of the driver who made the delivery
     * @param startTime When the delivery started
     * @param endTime When the delivery ended
     * @throws IllegalArgumentException if driver not found or times are invalid
     */
    public void recordDelivery(int driverId, LocalDateTime startTime, LocalDateTime endTime) {
        if (!drivers.containsKey(driverId)) {
            throw new IllegalArgumentException("Driver with ID " + driverId + " not found");
        }
        if (!endTime.isAfter(startTime)) {
            throw new IllegalArgumentException("End time must be after start time");
        }
        
        deliveries.add(new DeliveryRecord(driverId, startTime, endTime));
    }

    /**
     * Calculate total earnings for a specific driver.
     *
     * @param driverId ID of the driver
     * @return Total earnings in USD
     * @throws IllegalArgumentException if driver not found
     */
    public double getDriverEarnings(int driverId) {
        if (!drivers.containsKey(driverId)) {
            throw new IllegalArgumentException("Driver with ID " + driverId + " not found");
        }
        
        double hourlyRate = drivers.get(driverId);
        double totalHours = 0.0;
        
        for (DeliveryRecord delivery : deliveries) {
            if (delivery.getDriverId() == driverId) {
                double duration = java.time.Duration.between(
                    delivery.getStartTime(), 
                    delivery.getEndTime()
                ).toMinutes() / 60.0; // Convert to hours
                totalHours += duration;
            }
        }
        
        return totalHours * hourlyRate;
    }

    /**
     * Get all delivery records.
     *
     * @return List of delivery records
     */
    public List<DeliveryRecord> getAllDeliveries() {
        return new ArrayList<>(deliveries);
    }

    /**
     * Get all delivery records for a specific driver.
     *
     * @param driverId ID of the driver
     * @return List of delivery records for the driver
     * @throws IllegalArgumentException if driver not found
     */
    public List<DeliveryRecord> getDriverDeliveries(int driverId) {
        if (!drivers.containsKey(driverId)) {
            throw new IllegalArgumentException("Driver with ID " + driverId + " not found");
        }
        
        List<DeliveryRecord> driverDeliveries = new ArrayList<>();
        for (DeliveryRecord delivery : deliveries) {
            if (delivery.getDriverId() == driverId) {
                driverDeliveries.add(delivery);
            }
        }
        return driverDeliveries;
    }

    /**
     * Inner class to represent a delivery record
     */
    private static class DeliveryRecord {
        private final int driverId;
        private final LocalDateTime startTime;
        private final LocalDateTime endTime;

        public DeliveryRecord(int driverId, LocalDateTime startTime, LocalDateTime endTime) {
            this.driverId = driverId;
            this.startTime = startTime;
            this.endTime = endTime;
        }

        public int getDriverId() {
            return driverId;
        }

        public LocalDateTime getStartTime() {
            return startTime;
        }

        public LocalDateTime getEndTime() {
            return endTime;
        }
    }
} 

File: delivery.py
from datetime import datetime
from typing import Dict, List, Optional

class Delivery:
    def __init__(self):
        # Dictionary to store driver information: {driver_id: hourly_rate}
        self.drivers: Dict[int, float] = {}
        # List to store delivery records: [(driver_id, start_time, end_time)]
        self.deliveries: List[tuple] = []

    def add_driver(self, driver_id: int, usd_hourly_rate: float) -> None:
        """
        Add a new driver with their hourly rate.
        
        Args:
            driver_id (int): Unique identifier for the driver
            usd_hourly_rate (float): Driver's hourly rate in USD
        """
        if driver_id in self.drivers:
            raise ValueError(f"Driver with ID {driver_id} already exists")
        if usd_hourly_rate <= 0:
            raise ValueError("Hourly rate must be greater than 0")
        
        self.drivers[driver_id] = usd_hourly_rate

    def record_delivery(self, driver_id: int, start_time: datetime, end_time: datetime) -> None:
        """
        Record a delivery with its start and end times.
        
        Args:
            driver_id (int): ID of the driver who made the delivery
            start_time (datetime): When the delivery started
            end_time (datetime): When the delivery ended
        """
        if driver_id not in self.drivers:
            raise ValueError(f"Driver with ID {driver_id} not found")
        if end_time <= start_time:
            raise ValueError("End time must be after start time")
        
        self.deliveries.append((driver_id, start_time, end_time))

    def get_driver_earnings(self, driver_id: int) -> float:
        """
        Calculate total earnings for a specific driver.
        
        Args:
            driver_id (int): ID of the driver
            
        Returns:
            float: Total earnings in USD
        """
        if driver_id not in self.drivers:
            raise ValueError(f"Driver with ID {driver_id} not found")
        
        hourly_rate = self.drivers[driver_id]
        total_hours = 0.0
        
        for d_id, start, end in self.deliveries:
            if d_id == driver_id:
                duration = (end - start).total_seconds() / 3600  # Convert to hours
                total_hours += duration
        
        return total_hours * hourly_rate

    def get_all_deliveries(self) -> List[tuple]:
        """
        Get all delivery records.
        
        Returns:
            List[tuple]: List of (driver_id, start_time, end_time) tuples
        """
        return self.deliveries.copy()

    def get_driver_deliveries(self, driver_id: int) -> List[tuple]:
        """
        Get all delivery records for a specific driver.
        
        Args:
            driver_id (int): ID of the driver
            
        Returns:
            List[tuple]: List of (driver_id, start_time, end_time) tuples
        """
        if driver_id not in self.drivers:
            raise ValueError(f"Driver with ID {driver_id} not found")
        
        return [(d_id, start, end) for d_id, start, end in self.deliveries if d_id == driver_id] 